### PR TITLE
Binary Search in TimeZoneInfo.GetAdjustmentRuleForTime

### LINKED ISSUE #5716: Make search in TimeZoneInfo.GetAdjustmentRuleForTime more efficient
While investigating dotnet/coreclr#5558 it was noted that TimeZoneInfo.GetAdjustmentRuleForTime can be made more efficient, especially on Linux.

@eerhardt noted:

The Adjustment Rules between Windows and Linux are vastly different.

On Windows, for a given time zone, there are usually < 5.  And they contain general rules like:
-          Before 2006, DST starts on the first Sunday in April and ends on the last Sunday in October
-          After 2007, DST starts on the last Sunday in March and ends on the first Sunday in November

The Windows format is not precise enough to be accurate for all the different time zone rules in the world.

On Linux, the standard is to use the Olson tzdata files.  Their format is:
-          For a given time zone, every time the offset from UTC changes, make an entry
-          For something like the US, there will be one of these entries 2x for every year (one when DST turns on, one when off)

Both of these formats get turned into “AdjustmentRules”.  On Windows there will be only a handful.  On Linux there will be a lot.

For both platforms, finding the Adjustment rule starts at the beginning and ‘for’s through the list until it finds it.  For windows it is quick because there aren’t that many. On Linux, I can see it taking much longer.

I can imagine a good fix here would be to binary search the adjustment rules instead.  We have some good tests, so we would know pretty quick if something was broken.

@mconnew suggested an alternative approach: 
An optimization I plan on testing would be to maintain a year based index of starting points in the array. So if I have a date from 1985, it will return the index to the first potential matching rule, then the loop would simply start at that point. I presume the rules are already sorted in date order?

@mj1856 noted:
Given the size of the data for each individual time zone, I really don’t think there's going to be a considerable difference between a simple binary search and searching a precomputed index.  However, building the index will require a full pass through the data, and a small bit of memory to hold the index.  It's a classic trade-off.

Also consider that even if the index yield faster lookup time for bulk operations, it adds unnecessary latency for single-use operations, which can be just as common.

So while I can see how the index approach might work, I think it's not warranted yet.  Let's see first how a binary search performs.  It will probably be good enough.

### PR DESCRIPTION
**Pull request by eerhardt**
Since Unix uses IANA tzdata files, there can be hundreds of adjustment rules for each time zone.  To make the search for a given rule faster, use a binary search since the rules are guaranteed to be in chronological order.  See ValidateTimeZoneInfo.

Fix #5716 

@tarekgh @stephentoub @mellinoe 

/cc @erozenfeld

I ran the following program on my machines (note: 3 completely different physical machines):

```C#
using System;
using System.Diagnostics;

namespace TZPerfTest
{
    class Program
    {
        static void Main(string[] args)
        {
            DateTime dt;
            for (int i = 0; i < 20; i++)
            {
                Stopwatch watch = new Stopwatch();
                watch.Start();
                for (int j = 0; j < 10000; j++)
                {
                    dt = DateTime.Now; dt = DateTime.Now; dt = DateTime.Now;
                    dt = DateTime.Now; dt = DateTime.Now; dt = DateTime.Now;
                    dt = DateTime.Now; dt = DateTime.Now; dt = DateTime.Now;
                }

                Console.Write(watch.ElapsedMilliseconds + "<br>");
            }
        }
    }
}
```

| Platform | With My Change|Without My Change|
| -------- | :-------------------------------------: | :-------------------------------------: |
| **Windows x64** | 22<br>21<br>20<br>19<br>20<br>19<br>20<br>20<br>20<br>22<br>20<br>20<br>20<br>22<br>19<br>20<br>21<br>20<br>20<br>19<br> | 21<br>20<br>20<br>21<br>20<br>20<br>19<br>20<br>21<br>20<br>20<br>22<br>21<br>20<br>20<br>21<br>20<br>19<br>21<br>21<br> |
| **MacBook Pro** | 147<br>144<br>138<br>146<br>141<br>137<br>142<br>146<br>138<br>141<br>142<br>146<br>144<br>140<br>137<br>138<br>138<br>141<br>140<br>140<br> | 344<br>318<br>308<br>327<br>347<br>313<br>310<br>317<br>325<br>324<br>304<br>309<br>332<br>302<br>302<br>321<br>320<br>315<br>317<br>311<br> |
| **Ubuntu 14.04 VM** | 162<br>168<br>156<br>156<br>156<br>156<br>163<br>157<br>160<br>156<br>156<br>156<br>157<br>186<br>158<br>156<br>157<br>158<br>156<br>156<br> | 346<br>339<br>340<br>338<br>339<br>340<br>338<br>340<br>338<br>339<br>338<br>339<br>339<br>341<br>338<br>339<br>340<br>341<br>346<br>339<br> |

### CODE REVIEW

**Comment by eerhardt** [comment]
@dotnet-bot test Ubuntu x64 corefx_baseline
@dotnet-bot test Windows_NT x64 corefx_baseline

File: src/mscorlib/src/System/TimeZoneInfo.cs [hunk 1]
```diff
@@ -1062,14 +1062,29 @@ private AdjustmentRule GetAdjustmentRuleForTime(DateTime dateTime, bool dateTime
...
             {
-                AdjustmentRule rule = _adjustmentRules[i];
-                AdjustmentRule previousRule = i > 0 ? _adjustmentRules[i - 1] : rule;
-                if (IsAdjustmentRuleValid(rule, previousRule, dateTime, date, dateTimeisUtc))
+                int median = low + ((high - low) >> 1);
+
+                AdjustmentRule rule = _adjustmentRules[median];
+                AdjustmentRule previousRule = median > 0 ? _adjustmentRules[median - 1] : rule;
+
+                int compareResult = CompareAdjustmentRule(rule, previousRule, dateTime, date, dateTimeisUtc);
```
Comment (tarekgh) [inline src/mscorlib/src/System/TimeZoneInfo.cs]: >CompareAdjustmentRule [](start = 36, length = 21)

nit: the name CompareAdjustmentRule is kind of confusing that I read it comparing 2 adjustment rules together.

**Comment by tarekgh** [comment]
:+1:

File: src/mscorlib/src/System/TimeZoneInfo.cs (hunk 1 above)
Comment (eerhardt) [inline src/mscorlib/src/System/TimeZoneInfo.cs]: I can change it to `CompareAdjustmentRuleToDateTime`.  Thoughts?

**Comment by danmosemsft** [comment]
I'm curious roughly how Windows perf compares after this change.

**Comment by danmosemsft** [comment]
Are there still potenntial improvements for Unix that you can imagine? Since it's still 6x regressed from Windows. Maybe we should have a tracking issue.

**Comment by eerhardt** [comment]
> Are there still potenntial improvements for Unix that you can imagine? Since it's still 6x regressed from Windows. Maybe we should have a tracking issue.

I think there probably are, but I didn't investigate it.

We have https://github.com/dotnet/corefx/issues/3571 tracking the overall `DateTime.Now` perf scenario on Linux. I've left that open because of the big discrepancy I'm seeing on my machines.

**Comment by eerhardt** [comment]
test Windows_NT x64 Checked Build and Test (Jit - CoreFx)

I don't think I've broken the `System.Linq.Expressions.Tests`.  This leg isn't a required leg, so it may be broken.  https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x64_checked_windows_nt_corefx_baseline_prtest/50/consoleFull
```
15:18:11 D:\j\workspace\x64_checked_w---d7295605\_\fx\Tools\tests.targets(287,5): warning MSB3073: The command "D:\j\workspace\x64_checked_w---d7295605\_\fx\bin/AnyOS.AnyCPU.Release/System.Linq.Expressions.Tests/netcoreapp//RunTests.cmd D:\j\workspace\x64_checked_w---d7295605\_\fx\bin/testhost/netcoreapp-Windows_NT-Release-x64/" exited with code -1073741819. [D:\j\workspace\x64_checked_w---d7295605\_\fx\src\System.Linq.Expressions\tests\System.Linq.Expressions.Tests.csproj]
15:18:11 
D:\j\workspace\x64_checked_w---d7295605\_\fx\Tools\tests.targets(295,5): error : One or more tests failed while running tests from 'System.Linq.Expressions.Tests' please check D:\j\workspace\x64_checked_w---d7295605\_\fx\bin/AnyOS.AnyCPU.Release/System.Linq.Expressions.Tests/netcoreapp/testResults.xml for details! [D:\j\workspace\x64_checked_w---d7295605\_\fx\src\System.Linq.Expressions\tests\System.Linq.Expressions.Tests.csproj]
```

**Comment by eerhardt** [comment]
@dotnet-bot test Windows_NT x64 corefx_baseline

**Comment by danmosemsft** [comment]
@eerhardt that same error occurred when I ran the corefx tests in the coreclr CI yesterday. I believe there is some mismatch being exposed. Running locally, it passed sporadically. I belive you can go ahead and merge if @tarekgh is signed off.

**Comment by stephentoub** [comment]
LGTM. Thanks, @eerhardt.

**Pull request merged**